### DISCUSSION TITLE
My intuition vs. Uncertainties

### DISCUSSION THREAD

**Discussion by adriesse** [lmfit/uncertainties#177] [Q&A] [answered]
Ok, I'm sure my intuition would lose, but I'm hoping I can improve my intuition by playing with _uncertainties_...

I take voltage measurements that have uncertainty due to a potential zero offset as well as a potential gain error.  Typical scenario I would say.  The two sources of uncertainty are uncorrelated, but sequential measurements suffer the same errors and have the same uncertainties.

If I have two measurements and calculate their sum, _uncertainties_ appears to take a simple sum of their offset uncertainties and a simple sum of their gain uncertainties, and subsequently combines them (root-sum-square).

If I combine the two uncorrelated uncertainties for each measurement manually (root-sum-square) and subsequently take the simple sums I don't get the same combined uncertainty (unless the two measured values happen to be identical).

So the sequence of operations is important, but I can't think of any reason why one is better than the other.  Can anyone explain why the former sequence of operations (_uncertainties_) is correct and what is wrong with the latter?

**Comment by lebigot** [comment, marked as answer]
In order to reconcile your intuition with `uncertainties`, you need to make sure that the voltage is correctly modeled. In this case, we need two variables that model uncertainties (offset and gain). Each voltage is then offset + gain * real voltage. Summing them will do a mix of root-sum-squares and straight uncertainty sums.

If you use this, then your intuition is hopefully the same as what `uncertainties` gives.

If it isn't then it's best to provide equations and numerical values so that we can find the culprit.

**Reply by adriesse** [reply]
Hi Eric,

Here is a little code fragment.  The last two lines show the two options for order of operations with different results.  I can see _advantages_ to the first option, but I can't see what's fundamentally _wrong_ with the second option.
```
from uncertainties import ufloat

from numpy import sum, sqrt, square

def rss(a, axis=None):
    return sqrt(sum(square(a), axis))

gain = ufloat(1, 0.02, 'gain') # 2% uncertainty on gain
zero = ufloat(0, 5, 'zero') # 5 mV zero offset uncertainty

# two voltage measurements of 10 and 200 mV respectively
v1 = 10 * gain + zero # uncertainties 0.2 and 5 mV
v2 = 200 * gain + zero # uncertainties 4.0 and 5 mV

# combined uncertainties of individual measurements
print(v1.n, v1.s, '==', rss([0.2, 5.0]))
print(v2.n, v2.s, '==', rss([4.0, 5.0]))

v3 = v1 + v2

# uncertainties appears to sum the gain errors and the offset errors
# and then combines them via rss
print(v3.n, v3.s, '==', rss([sum([0.2, 4.0]), sum([5.0, 5.0])]) )
print(v3.n, v3.s, '!=', sum([rss([0.2, 5.0]), rss([4.0, 5.0])]) )


```

**Reply by lebigot** [reply]
Thank you for the precisions.

There is something fundamentally wrong with the second option: it is simply a formal play with formulas and is not grounded in any theory, if I'm not mistaken. 😉

The first option is correct because, say, `zero+zero` = `2*zero` is indeed equal to 0±10 mv, so you directly sum the two 5 mV uncertainties.

Now, there is a third option that could make sense, depending on how you want to model the voltmeter: if each time you measure a voltage the gain and the zero offset change, then they are _independent_ of the gain and offset of the previous measurement. You would express this by creating a new `gain` and a new `zero` random variable each time you "do" a measurement:

```python
from copy import copy

v1 = 10 * gain + zero
v2 = 200 * copy(gain) + copy(zero)  # Creation of new random, *independent* gain and zero offset
```

In this case, I am expecting an uncertainty of `rss([rss([0.2, 5.0]), rss([4.0, 5.0])])` = `rss([0.2, 5.0, 4.0, 5.0])` since we have 4 independent random variables.

Does this clarify things enough for you?

**Reply by adriesse** [reply]
Yes, I understand the third option. And yes, I am playing around with the formulas rather than studying theories!

But it is not just fun and games because the situation arises when I receive a series of measurements (or some other input quantity) each with a pre-combined uncertainty rather than the component uncertainties. Then the only calculation I can carry out is the second option.

Or sometimes I can beg the supplier for more information. In these cases it would be nice to have a short, intuitive statement to explain why having their face values of the uncertainties is not enough but I actually have to look inside them.

So far this discussion is certainly helpful in confirming that I am doing the right thing, and much appreciated. And maybe there is no intuitive explanation possible. I guess as a last resort I could point the person to this discussion on github. :)

**Reply by lebigot** [reply]
I can feel the pain.

Now, I'm not sure about what intuition you're after, but if you give more details, maybe I can help.

In general, what is key is to understand the concept of random variable (e.g., I take a voltmeter, another one, another one, etc. and do statistics on the voltage they display when shorting the electrodes, and this gives the `zero` offset with its uncertainty).

Then what matters is to understand correlations. For example, whatever the value of `zero`, `zero+zero` is always equal to `2*zero` and therefore its uncertainty is _double_ that of `zero` (so we don't do the root-sum-square, which would give $\sqrt 2$), and similarly for `zero-zero`, which is always equal to 0 (while the root-sum-square still gives a $\sqrt 2$ factor in the uncertainty compared to that of `zero`). Only fully independent variables have a sum whose uncertainty can always be obtained with the root-sum-square formula.

**Reply by adriesse** [reply]
Thanks for your offer!  I think I'm good with the basics and perhaps I can substitute faith for intuition in this particular matter. 

There are much harder things also, like sources of uncertainty that are not actually random.  But that will be for another day!
